### PR TITLE
Fix a De Bruijn bug in the computation of term relevance in the kernel.

### PR DESCRIPTION
**Pull request by ppedrot**
Opening up a lambda should always lift the substitution attached to it. You can probably weaponize this to a proof of False.

### CODE REVIEW

**Comment by JasonGross** [comment]
This should probably get a new test-case in the test-suite, right?

**Comment by ppedrot** [comment]
The file I was writing was fairly big and relying on UIP, I don't know whether I can generate a smaller, self-contained example.

**Comment by SkySkimmer** [comment]
~~~coq
Definition a := fun (P:SProp) (p:P) => p.

Lemma foo : (let k := a in let k' := a in fun (x:nat) y => x) = (let k := a in fun x y => y).
Proof.
reflexivity.
Qed.

Lemma oops : False.
Proof.
  pose proof (f_equal (fun F => F 0 1) foo : 0 = 1) as e.
  inversion e.
Qed.
Print Assumptions oops.
~~~

**Review by JasonGross — DISMISSED** [review]
Thanks, @skyskimmer!  This should definitely get a test-suite file and an entry in the list of critical bugs.

Does this affect the checker too, or only coqc?

**Comment by SkySkimmer** [comment]
Checker too

Gaëtan Gilbert

**Comment by ppedrot** [comment]
@SkySkimmer feel free to push your example atop of this PR.

File: dev/doc/critical-bugs [hunk 1]
```diff
@@ -250,6 +250,17 @@ Conversion machines
...
 
+  component: lazy machine
+  summary: incorrect De Bruijn handling when inferring the relevance mark for a lambda
+  introduced: 2019-03-15, 23f84f37c674a07e925925b7e0d50d7ee8414093 and 71b9ad8526155020c8451dd326a52e391a9a8585, SkySkimmer
+  impacted released versions: 8.10.0
+  impacted coqchk versions: 8.10.0
+  found by: ppedrot investigating unexpected conversion failures with SProp
+  exploit: test-suite/bugs/closed/bug_10904.v
+  GH issue number: #10904
+  risk: none without using -allow-sprop (off by default in 8.10.0),
```
Comment (gares) [inline dev/doc/critical-bugs]: It this true? didn't we enable it in the end?

Comment (SkySkimmer) [reply dev/doc/critical-bugs]: https://github.com/coq/coq/pull/10811 merged 2 days ago
Coq 8.10.0 8 days ago

Comment (gares) [reply dev/doc/critical-bugs]: So I guess this bug blocks an eventual backport of #10811 for 8.10.1 (not sure it was planned, but in case).

Comment (Zimmi48) [reply dev/doc/critical-bugs]: Such a PR was definitely out of scope for a patch-level release in any case.

File: doc/changelog/01-kernel/10904-fix-debruijn-sprop-rel.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+- Fix proof of False when using |SProp| (incorrect De Bruijn handling
```
Comment (Zimmi48) [inline doc/changelog/01-kernel/10904-fix-debruijn-sprop-rel.rst]: `|SProp|` is not defined:

```suggestion
- Fix proof of False when using :math:`\SProp` (incorrect De Bruijn handling
```

**Review by Zimmi48 — APPROVED** [review]
Adding "needs: fixing" label for failure on doc.

File: doc/changelog/01-kernel/10904-fix-debruijn-sprop-rel.rst (hunk 2 above)
Comment (SkySkimmer) [reply doc/changelog/01-kernel/10904-fix-debruijn-sprop-rel.rst]: I'm defining it instead.

**Comment by ppedrot** [comment]
AFAICT this is ready and solely missing an assignee.

**Review by SkySkimmer — APPROVED** [review]
A kernel oops from me

**Pull request merged**